components/ide/jetbrains/backend-plugin/src/main/kotlin/io/gitpod/jetbrains/remote/GitpodCLIService.kt

```diff
@@ -53,22 +54,31 @@ class GitpodCLIService : RestService() {
         return "invalid operation"
     }
 
-    private fun withClient(request: FullHttpRequest, context: ChannelHandlerContext, action: (project: Project?) -> Unit): String? {
+    private tailrec fun getClientSessionAsync(): ClientSession {
         val project = getLastFocusedOrOpenedProject()
         var session: ClientSession? = null
         if (project != null) {
-            session = ClientSessionsManager.getProjectSessions(project, false).first()
+            session = ClientSessionsManager.getProjectSessions(project, false).firstOrNull()
         }
         if (session == null) {
-            session = ClientSessionsManager.getAppSessions(false).first()
+            session = ClientSessionsManager.getAppSessions(false).firstOrNull()
         }
-        if (session == null) {
-            return "no client"
+        return if (session != null) {
+            session
+        } else {
+            Thread.sleep(1000L)
+            getClientSessionAsync()
         }
-        ClientId.withClientId(session.clientId) {
-            action(project)
+    }
+
+    private fun withClient(request: FullHttpRequest, context: ChannelHandlerContext, action: (project: Project?) -> Unit): String? {
+        ApplicationManager.getApplication().executeOnPooledThread {
+            val session = getClientSessionAsync()
+            ClientId.withClientId(session.clientId) {
+                action(getLastFocusedOrOpenedProject())
```

Comment:
you should pass here a project from getClientSessionAsync

Reply:
Updated!